### ISSUE TITLE
Replace FILLVAL with np.nan in read_cdf()

### ISSUE THREAD

**Issue by jgieseler** [sunpy/sunpy#5908]
### Describe the feature

Previously, Heliopy replaced the `FILLVAL` defined for a variable in a CDF file with `np.nan` (cf. https://github.com/heliopython/heliopy/blob/0.15.4/heliopy/data/util.py#L825). Is there a reason why this is not done in SunPy's `read_cdf()`? (Probably @dstansby)

### Proposed solution

Adopt old Heliopy code from https://github.com/heliopython/heliopy/blob/0.15.4/heliopy/data/util.py#L825

**Comment by dstansby** [comment]
No reason, other than I probably forgot. We should definitely do this.

**Comment by dstansby** [comment]
Had a quick think about this, and instead of filling with `NaN` we should try and used masked arrays and set the mask. That way we're not limited to applying fillval to just floating point columns.

**Comment by jgieseler** [comment]
Yes, that would solve the problem of assigning `np.nan` to an integer column. I think I can take care of it.

**Comment by dstansby** [comment]
I'm just working on this, but it doesn't look like our current CDF test file has any masked values. @jgieseler can you recommend a small (in filesize) CDF file that has some FILLVAL values that we can use to test with?

**Comment by dstansby** [comment]
Just coming back to this, and sadly it looks like pandas doesn't have support for masking data 😢 However, once https://github.com/sunpy/sunpy/pull/5834 is merged we can mask the values in the astropy Timeseries (see https://docs.astropy.org/en/stable/timeseries/masking.html)

**Comment by dstansby** [comment]
The issue is that `NaN` is a floating point number, and there's no equivalent for integer data types. In `heliopy` I hacked this by converting all integer data type columns to floating point first, but that was a Bad Idea because it turns out floating point numbers take up much more space than integers (who knew?!). So the approach in that PR works fine for floating point columns, but I'd like it to work more generally which requires support for properley masking data.

**Comment by dstansby** [comment]
I'm going to re-open this as we only fixed it for floating point data types.

**Comment by nabobalis** [comment]
Do the integer data have fill values too?

If so, we shouldn't need to worry about DTYPE and just replace it wholesale.

**Comment by dstansby** [comment]
They do, but `NaN` is a floating point number, with no integer equivalent, so we can't set the masked values to `NaN` for integer data.

**Comment by nabobalis** [comment]
They don't have a seperate fill value for INT data stored in the metadata?

**Comment by dstansby** [comment]
They do, but the point is to mask out that value so for e.g. plotting it doesn't show up.

**Comment by jeandet** [comment]
My 2 cents here, one benefit with NaN is that most plotting libraries handles them as line break, with masked arrays, you won't easily see that those points are missing.

**Comment by dstansby** [comment]
`pandas` now has a [nullable integer data type](https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html) (which is basically NaN for ints). I forget if the relevant code here using pandas underneath, but if it does replacing the fillval with a null value for integer data types might now be possible.

**Comment by jgieseler** [comment]
> `pandas` now has a [nullable integer data type](https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html) (which is basically NaN for ints). I forget if the relevant code here using pandas underneath, but if it does replacing the fillval with a null value for integer data types might now be possible.

That looks promising! I just checked `read_cdf()`, and in principle it should only require one line of code to convert an integer array to this specific Pandas [IntegerArray](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.arrays.IntegerArray.html#pandas.arrays.IntegerArray) (see https://github.com/jgieseler/sunpy/commit/30bd7d30eb4e03178681e41c6cbde80ac14cc9b8).

The question is if we want to use Pandas [IntegerArray](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.arrays.IntegerArray.html#pandas.arrays.IntegerArray) as long as it is flagged _experimental_ ("IntegerArray is currently experimental, and its API or internal implementation may change without warning.")? 🤔

**Comment by jgieseler** [comment]
**Addendum:** For multidimensional data it gets a little bit more tricky because Pandas [IntegerArray](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.arrays.IntegerArray.html#pandas.arrays.IntegerArray) only supports 1D data (as of now). So my workaround is to do the FILLVAL-replacement later, after the multidimensional data has been loaded into a dictionary of 1D arrays with column names such as `var_key_1`, see https://github.com/jgieseler/sunpy/commit/d808f097034fbea0d52f4474604c73a1f607cc9d
